guide.md

```diff
@@ -0,0 +1,559 @@
+CC toolchain available. This means it's no longer possible to unintentionally
+use `@local_config_cc`, as that toolchain won't be able to find the global
+binaries it wants and will fail to work.
+
+To enter the development shell, one can run:
+```
+$ nix develop
+```
+
+When running this for the first time, a `flake.lock` file is initialised,
+containing the exact versions of each dependency in `inputs` (which for us is
+basically just `nixpkgs`). This can be committed to the repository, and ensures
+that when others use the same environment, they will be provided the same
+versions of those dependencies, which means that the environment will be
+consistent across platforms. We will also use that same `flake.lock` as the
+single source of truth for the Nixpkgs version used in our build below.
+
+It's worth explicitly mentioning how useful being able to fix a version of
+Nixpkgs is for us: doing this allows us to select a consistent set of package
+revisions in one go, instead of having to maintain dependency versions for
+individual dependencies. Often one prefers to choose a tag for a stable release,
+which are of the form `nixos-yy.mm` where `yy` are the last two digits of a
+year, and `mm` is a month. Each stable release tag undergoes stringent
+verification via CI that packages in it will function correctly individually and
+play well together.
+
+### Upgrading the Nixpkgs version used (and other flake inputs)
+
+As with all dependencies, it's good to upgrade the version of the `nixpkgs`
+input from time to time, which will bump the versions of all Nix-provided
+dependencies. In the example above, the `nixpkgs` dependency is set to follow
+the `nixos-22.05` tag, and we can pull the latest commit from that branch using:
+
+```
+$ nix flake lock --update-input nixpkgs
+```
+
+It is also possible to change the Nixpkgs dependency to any branch: for
+instance, we could switch to `master` and receive bleeding-edge dependencies, or
+(more practically) we could choose to follow a newer tag (22.11 is available
+right now) by making the following change in `flake.nix`:
+
+```diff
+-    nixpkgs.url = "github:NixOS/nixpkgs/nixos-22.05";
++    nixpkgs.url = "github:NixOS/nixpkgs/nixos-22.11";
+```
+
+To update _all_ dependencies, one uses:
+
+```
+$ nix flake update
+```
+
+In the current example, this doesn't change much, since the only dependency we
+have apart from Nixpkgs itself is `flake-utils`, which doesn't change much.
+However, if you integrate more Nix-based dependencies or utilities into your
+shell, this will take care of all of them in one go, and update the `flake.lock`
+to include all of the changes.
+
+## Using `direnv` with Nix
+
+The use of this feature can be made even more seamless by using
+[direnv](direnv), which has support for the `flake.nix` file we used to describe
+the development shell.
+
+Put the following in a file called `.envrc` at the root of the repo:
```

Comment:
The `.envrc` file is currently missing in `examples/cc-template`. Was the intention to include it?

Reply:
Yep, thanks for the catch. (`.envrc` is in my global gitignore.)